src/runtime: second draft of map growth code

dgryski opened the pull request:
Fixes #1553

dgryski committed:
src/runtime: first darft of map growth code

Fixes #1553

dgryski committed:
src/runtime/hashmap: comments for iterator structure

dgryski committed:
testdata: add test for mapgrowth logic

dgryski commented:
~For some reason, tinygo/llvm didn't like the `n := *m` and `*m = n` lines from my original patch.  That points to a larger issue that needs to be fixed, but this solves the failing `hash/adler32` test.~

Edit: nope, this was totally incorrect.

dgryski committed:
src/runtime: prevent overflow when calculating hashmap growth limits

dgryski committed:
src/runtime: remove extra if check in hashmap set logic

dgryski commented:
Still have some `uintptr` potential overflow issues to solve.

dgryski committed:
src/runtime: make hashmap calculations more uintptr-size independent

deadprogram commented:
It is all passing and looks good to my little eye. @aykevl any feedback here?

aykevl commented on src/runtime/hashmap.go:
```diff
@@ -57,12 +68,70 @@ func hashmapMake(keySize, valueSize uint8, sizeHint uintptr) *hashmap {
+
+func hashmapShouldGrow(m *hashmap) bool {
+
+	if heapStart == heapEnd {
+		// don't try to grow if we have no memory
+		// this also triggers during interp
+		return false
 	}
```
I don't agree with this. I'm pretty sure this indicates a bug somewhere else in the code.
(I need to review the rest of the code still, it's probably fine)

dgryski replied:
Yeah, this was the workaround I found. I don't like it either.

deadprogram replied:
OK, so what do we do now? If this indicates a bug elsewhere, is there a way we can trigger it and find it?

If we are not going to do this right now, we could create an issue to track doing so, and then merge this PR (pending any other feedback) in the interest of unblocking this feature.

aykevl replied:
I have a strong suspicion this is an interp bug. One possible solution (that's still a hack but a lot cleaner) is to add `runtime.hashmapGrow` to the functions that are excluded from interp:

https://github.com/tinygo-org/tinygo/blob/9eb4a6268aa8d61de9d6c03d6d02c0f0c2be157f/interp/interpreter.go#L204-L217

aykevl replied:
```
heapStart: 0x0009fff0
heapEnd: 0x000a0000
```
Actually, it might be a GC issue after all. These two addresses are awkwardly close together.

aykevl replied:
I think I have a fix. I'll test it a bit more to confirm.

aykevl replied:
Here is the fix: https://github.com/tinygo-org/tinygo/pull/2812

deadprogram replied:
@dgryski #2812 is in `dev` now can you please rebase and make suitable adjustments to this PR?

deadprogram commented:
OK, seems like the main issue with this PR has now been addressed. Are we good to merge?

aykevl commented on main_test.go:
```diff
@@ -57,6 +57,7 @@ func TestBuild(t *testing.T) {
 		"json.go",
 		"map.go",
 		"math.go",
+		"mapgrowth.go",
```
Please keep this list sorted.
Also, why not merge mapgrowth.go with map.go? Unfortunately, compilations are still relatively slow and adding new tests slows down everything a lot.

dgryski replied:
It's a different file because I just added the test program I was using for development and dropped it into the testdata directory.  I'll add it to the `map.go` file properly.

aykevl commented on src/runtime/hashmap.go:
```diff
@@ -57,14 +68,65 @@ func hashmapMake(keySize, valueSize uint8, sizeHint uintptr) *hashmap {
+	case hashmapAlgorithmInterface:
+		return hashmapInterfaceEqual
+	default:
+		panic("unknown hashmap equal algorithm")
```
If we get here, that's a compiler bug. It should never ever be reached. I think returning nil is the better option to save binary size (or even something like `__builtin_unreachable()` in C, which isn't yet exposed in TinyGo).

aykevl commented on src/runtime/hashmap.go:
```diff
@@ -57,14 +68,65 @@ func hashmapMake(keySize, valueSize uint8, sizeHint uintptr) *hashmap {
+	case hashmapAlgorithmInterface:
+		return hashmapInterfacePtrHash
+	default:
+		panic("unknown hashmap hash algorithm")
```
Same here.

aykevl reviewed (commented):
This looks pretty reasonable. Just a few nits below.
There is an impact on code size, but in many cases it's not too big (especially considering this adds a new feature). And it may be avoidable by using the function pointers in a slightly different way.

dgryski committed:
src/runtime: return a nil pointer for compiler bugs in hashmap code

We'll still panic if there's a compiler bug, but at least we'll have smaller
code in all the cases where we don't.

dgryski committed:
testdata: move map growth test to map.go

dgryski commented:
PTAL (while CI churns...)

aykevl commented on src/runtime/hashmap.go:
```diff
@@ -57,14 +68,67 @@ func hashmapMake(keySize, valueSize uint8, sizeHint uintptr) *hashmap {
+	case hashmapAlgorithmInterface:
+		return hashmapInterfaceEqual
+	default:
+		// compiler bug :(
+		return nil
```
Technically correct. I think I would have called it "unreachable" but :shrug:

dgryski replied:
I'll fix that in a follow up PR.

deadprogram commented:
At long last merging! Fantastic work @dgryski
